vulkan: Update all components to Vulkan SDK 1.3.283.0

akien-mga opened the pull request:
Draft as I need help to rediff the SPIRV-Reflect specialization constant patch. CC @godotengine/rendering 

You can help me by checking out this PR, trying to apply `thirdparty/spirv-reflect/patches/specialization-constants.patch` with `patch -p3 < thirdparty/spirv-reflect/patches/specialization-constants.patch`, and then figuring out how to resolve the hunks that couldn't be applied. Once the changes are good, you can do `git diff > mydiff` and copy that file on top of `thirdparty/spirv-reflect/patches/specialization-constants.patch`, to include in the commit.

It seems like upstream added a (more minimal?) spec constant struct with the same name, so we likely need to fold our changes into their new API:
- https://github.com/KhronosGroup/SPIRV-Reflect/pull/242

Worth noting that there are two open PRs upstream to add the kind of spec constants support we have, which we could help drive to completion / merge:
- https://github.com/KhronosGroup/SPIRV-Reflect/pull/154 (derived from @reduz's https://github.com/godotengine/godot/pull/50325)
- https://github.com/KhronosGroup/SPIRV-Reflect/pull/197 (which was made independently later on, missing the fact there was already a PR for that)

akien-mga committed:
vulkan: Update all components to Vulkan SDK 1.3.183.0

Pass `VMA_ALLOCATOR_CREATE_KHR_MAINTENANCE5_BIT` to VMA when using Vulkan 1.3
features.

Co-authored-by: Pedro J. Estébanez <pedrojrulez@gmail.com>

RandomShaper commented:
Re-done patch to spirv-reflect:
[specialization-constants.patch](https://github.com/godotengine/godot/files/15333204/specialization-constants.patch)

akien-mga commented:
> Re-done patch to spirv-reflect: [specialization-constants.patch](https://github.com/godotengine/godot/files/15333204/specialization-constants.patch)

Thanks, added to the PR. Should be good to go then.

akien-mga commented:
@bruvzg The macOS and iOS builds for glslang are failing with:
```
In file included from thirdparty/glslang/glslang/GenericCodeGen/Link.cpp:40:
In file included from thirdparty/glslang/glslang/GenericCodeGen/../Include/ShHandle.h:48:
thirdparty/glslang/glslang/GenericCodeGen/../Include/InfoSink.h:104:37In file included from thirdparty/glslang/glslang/GenericCodeGen/CodeGen.cpp:36:
In file included from thirdparty/glslang/glslang/GenericCodeGen/../Include/ShHandle.h:48:
thirdparty/glslang/glslang/GenericCodeGen/../Include/InfoSink.h:104:37: error: 'absolute' is unavailable: introduced in macOS 10.15
            append(std::filesystem::absolute(shaderFileName).string());
                                    ^
: error: 'absolute' is unavailable: introduced in macOS 10.15
/Applications/Xcode_15.0.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__filesystem/operations.h:65:35: note: 'absolute' has been explicitly marked unavailable here
inline _LIBCPP_HIDE_FROM_ABI path absolute(const path& __p) { return __absolute(__p); }
                                  ^
```
```
In file included from thirdparty/glslang/glslang/GenericCodeGen/Link.cpp:40:
In file included from thirdparty/glslang/glslang/GenericCodeGen/../Include/ShHandle.h:48:
thirdparty/glslang/glslang/GenericCodeGen/../Include/InfoSink.h:104:37: error: 'absolute' is unavailable: introduced in iOS 13.0 [2]
             append(std::filesystem::absolute(shaderFileName).string());
                                     ^
```

Sounds like using this glslang version would force us to bump our min macOS and iOS releases.
Should we send a patch upstream to restore support for older versions?

bruvzg commented:
> Sounds like using this glslang version would force us to bump our min macOS and iOS releases.
> Should we send a patch upstream to restore support for older versions?

Vulkan (with a feature set used by Godot) won't properly work on macOS 10.14 and iOS 12, it's only for OpenGL backend. Both versions are unsupported for a long time (but current Xcode still can target both), so I'm not sure if there's much reason to keep supporting it, or bump the min versions.

akien-mga commented:
> > Sounds like using this glslang version would force us to bump our min macOS and iOS releases.
> > Should we send a patch upstream to restore support for older versions?
> 
> Vulkan (with a feature set used by Godot) won't properly work on macOS 10.14 and iOS 12, it's only for OpenGL backend. Both versions are unsupported for a long time (but current Xcode still can target both), so I'm not sure if there's much reason to keep supporting it, or bump the min versions.

It probably doesn't make sense to attempt using our Vulkan/Metal renderers on macOS 10.14 or iOS 12 indeed, but for OpenGL I think it may still make sense if we can preserve compatibility.

Judging by https://gs.statcounter.com/ios-version-market-share/ stats, iOS 12.5 (which seemed to be a LTS and only got EOL'ed in Jan 2023) still has around 1.5% market share among iOS users:
![image](https://github.com/godotengine/godot/assets/4701338/d23457d9-1eb1-4dfc-b6c3-0be906b9075e)

bruvzg commented:
Seems like it only for output prints with the specific flag, so we can patch it locally by removing `if (absolute)` condition.

akien-mga commented on thirdparty/glslang/glslang/Include/InfoSink.h:
```diff
@@ -94,11 +95,22 @@ class TInfoSinkBase {
         snprintf(locText, maxSize, ":%d", loc.line);
-        append(loc.getStringNameOrNum(false).c_str());
+
+        if(loc.getFilename() == nullptr && shaderFileName != nullptr && absolute) {
+            //append(std::filesystem::absolute(shaderFileName).string());
+        } else {
+            std::string location = loc.getStringNameOrNum(false);
+            //if (absolute) {
+            //    append(std::filesystem::absolute(location).string());
+            //} else {
+                append(location);
+            //}
+        }
```
Made this small hack to restore compat for older Apple platforms, until the issue is resolved upstream or we decide to increase our min supported versions to drop iOS 12.0 and macOS 10.13 and 10.14.

bruvzg commented:
> For macOS, stats from the same website

That's some strange data, seem to include only outdated and unsupported versions of macOS, currently only Monterey (12) and newer are getting security updates.

akien-mga commented:
> > For macOS, stats from the same website
> 
> That's some strange data, seem to include only outdated and unsupported versions of macOS, currently only Monterey (12) and newer are getting security updates.

Yeah I think they simple didn't update their heuristics when macOS bumped from 10.15 to 11.0, so they seem to track all >= 10.15 into "Catalina". The data for older releases may still be somewhat reliable.

akien-mga commented:
I just rebased, and updated VMA to the newly tagged 3.1.0 release (no change compared to previous git snapshot in that PR). So I assume the CI will still fail similarly.

Then I'll try reverting VMA to the previous version (that we have in the `master` branch) to see if that's indeed what causes our issue.

akien-mga commented:
So reverting VMA to an older commit fixed it, i.e. this commit (which I'll now remove): https://github.com/godotengine/godot/pull/92010/commits/2b466bd5b1eb50de699f0d75b38f325a940281f4

The error on CI is:
```
godot.linuxbsd.editor.dev.double.x86_64.san: thirdparty/vulkan/vk_mem_alloc.h:3931: bool FindMemoryPreferences(bool, const VmaAllocationCreateInfo&, VmaBufferImageUsage, VkMemoryPropertyFlags&, VkMemoryPropertyFlags&, VkMemoryPropertyFlags&): Assertion `0 && "VMA_MEMORY_USAGE_AUTO* values can only be used with functions like vmaCreateBuffer, vmaCreateImage so that the details of the created resource are known." " Maybe you use VkBufferUsageFlags2CreateInfoKHR but forgot to use VMA_ALLOCATOR_CREATE_KHR_MAINTENANCE5_BIT?"' failed.
```

And that's consistent with the fact that the VMA update mostly adds support for `KHR_MAINTENANCE5`.

I don't know much about that, but maybe @DarioSamo or @RandomShaper could have a look at what we're missing?

RandomShaper commented:
This patch may do the trick:
```
diff --git a/drivers/vulkan/rendering_device_driver_vulkan.cpp b/drivers/vulkan/rendering_device_driver_vulkan.cpp
index 131e0e4a8a..d38eb5a88d 100644
--- a/drivers/vulkan/rendering_device_driver_vulkan.cpp
+++ b/drivers/vulkan/rendering_device_driver_vulkan.cpp
@@ -942,6 +942,10 @@ Error RenderingDeviceDriverVulkan::_initialize_allocator() {
 	allocator_info.physicalDevice = physical_device;
 	allocator_info.device = vk_device;
 	allocator_info.instance = context_driver->instance_get();
+	const bool use_1_3_features = physical_device_properties.apiVersion >= VK_API_VERSION_1_3;
+	if (use_1_3_features) {
+		allocator_info.flags |= VMA_ALLOCATOR_CREATE_KHR_MAINTENANCE5_BIT;
+	}
 	VkResult err = vmaCreateAllocator(&allocator_info, &allocator);
 	ERR_FAIL_COND_V_MSG(err, ERR_CANT_CREATE, "vmaCreateAllocator failed with error " + itos(err) + ".");
```

Explanation: with Vulkan >= 1.3, `KHR_MAINTENANCE5` extension is implicitly enabled. VMA needs to know if that's the case so it uses the right code path in the algo to find the right memory type.

Not tested, so, well, TIWAGOS.

akien-mga commented:
I gave it a try in the latest commit (7479f84), but that doesn't seem to solve it, it still fails at the same API test:
```
RenderSceneBuffersRD.create_texture_from_format --- executing with 5 parameters [&"", &"", <RDTextureFormat#-9223371657404467237>, <RDTextureView#-9223371657387690020>, false]
GDSCRIPT CODE:     RenderSceneBuffersRD.new().create_texture_from_format(StringName(""), StringName(""), RDTextureFormat.new(), RDTextureView.new(), false)
godot.linuxbsd.editor.dev.double.x86_64.san: thirdparty/vulkan/vk_mem_alloc.h:3931: bool FindMemoryPreferences(bool, const VmaAllocationCreateInfo&, VmaBufferImageUsage, VkMemoryPropertyFlags&, VkMemoryPropertyFlags&, VkMemoryPropertyFlags&): Assertion `0 && "VMA_MEMORY_USAGE_AUTO* values can only be used with functions like vmaCreateBuffer, vmaCreateImage so that the details of the created resource are known." " Maybe you use VkBufferUsageFlags2CreateInfoKHR but forgot to use VMA_ALLOCATOR_CREATE_KHR_MAINTENANCE5_BIT?"' failed.
```

RandomShaper commented:
I can't reproduce on Windows. However, I've found that this is something else we'd want to do when initializing `allocator_info`:
`allocator_info.vulkanApiVersion = physical_device_properties.apiVersion;`

It stops VMA assuming we're using Vulkan 1.0, and so lets it collect pointers to certain functions that make some operations more efficient.

Can you test it and see if it interacts with the issue at hand?

RandomShaper commented:
That's been useful! The problem is that piece of script is trying to create a texture with not a single usage bit specified. See https://github.com/godotengine/godot/pull/92587.

RandomShaper commented:
> That did the trick! Should I keep the changes we made to `allocator_info`, or revert them for now?

Good question! The one in the patch is needed as part of the update. However, the other line of code is not strictly part of the scope of this PR. It may be better to keep it as a separate commit. Besides, it's good to have the ability to frame that other change in a bisect.
